### PR TITLE
Refactor optimizer

### PR DESCRIPTION
**Pull request by FCO**
It's giving a error on this test and I don't know why...

```
$ perl6 -Ilib t/10-alternate-relation-modules.t
===SORRY!===
Incompatible MROs in P6opaque rebless for types ContainerDescriptor and Stash
```

### CODE REVIEW

**Comment by jonathanstowe** [comment]
I'm not quite seeing *how* but is it possible that ```::?CLASS``` has somehow wound up as the wrong thing?   And would it work if the ```::?CLASS.bless``` is  changed  to ```self.bless```?

**Comment by FCO** [comment]
@jonathanstowe I've tried that and also `self.WHAT.bless` but haven't change the error... :(

**Comment by FCO** [comment]
Oh! It passed!!!

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by FCO** [comment]
It haven't worked on my machine, but I've rebuilt my rakudo for the latest version, and that worked...

**Comment by FCO** [comment]
Could anyone take another look at it, please?

File: lib/Red/AST/Optimizer/AND.pm6 [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
 use Red::AST::Value;
+
 unit role Red::AST::Optimizer::AND;
 
 #| x > 1 AND x > 10 ==> x > 10
-multi method optimization-col1(
+multi method optimize(
     $left  where Red::AST::Ge|Red::AST::Gt,
-    $right where Red::AST::Ge|Red::AST::Gt
+    $right where Red::AST::Ge|Red::AST::Gt,
```
Comment (jonathanstowe) [inline lib/Red/AST/Optimizer/AND.pm6]: It looks like you've typed some of these often enough to warrant a  subset or two :-)

File: lib/Red/AST/Optimizer/AND.pm6 [hunk 2]
```diff
@@ -36,52 +39,60 @@ multi method optimization-col1(
...
     return ast-value False if $lv.defined and $rv.defined and $lv < $rv
 }
 
-#| x < 1 AND NOT(x >= 1) ==> True
-multi method optimization-col1(
+#| a.b AND NOT(a.b) ==> True
+multi method optimize(
     $left  where Red::Column,
-    $right where Red::AST::Not
+    $right where Red::AST::Not,
```
Comment (jonathanstowe) [inline lib/Red/AST/Optimizer/AND.pm6]: I'd go with a regular parameter declaration for these as as the ```where``` makes me think there is something special going on whereas there isn't.

File: lib/Red/AST/Optimizer/OR.pm6 [hunk 3]
```diff
@@ -0,0 +1,96 @@
+use Red::AST;
+use Red::AST::Infixes;
+use Red::AST::Value;
+unit role Red::AST::Optimizer::OR;
+
+#| x > 1 OR x > 10 ==> x > 10
+multi method optimize(
+    $left  where Red::AST::Ge|Red::AST::Gt,
```
Comment (jonathanstowe) [inline lib/Red/AST/Optimizer/OR.pm6]: I take back the subset from above, I'd create a (possibly empty) role for each set of AST types you declare like this.

Comment (FCO) [reply lib/Red/AST/Optimizer/OR.pm6]: I prefer to do not leave optimizations info leak for the outside world...

**Pull request merged**